source/cppfs/source/posix/LocalFileWatcher.cpp

```diff
@@ -0,0 +1,147 @@
+, m_fs(fs)
+, m_inotify(-1)
+{
+    // Create inotify instance
+    m_inotify = inotify_init();
+}
+
+LocalFileWatcher::~LocalFileWatcher()
+{
+    // Close watch handles
+    for (auto it : m_watchers) {
+        inotify_rm_watch(m_inotify, it.first);
+    }
+
+    // Close inotify instance
+    close(m_inotify);
+}
+
+AbstractFileSystem * LocalFileWatcher::fs() const
+{
+    // Return file system
+    return static_cast<AbstractFileSystem *>(m_fs.get());
+}
+
+void LocalFileWatcher::add(FileHandle & fh, unsigned int events, RecursiveMode recursive)
+{
+    // Get watch mode
+    uint32_t flags = 0;
+    if (events & FileCreated)  flags |= IN_CREATE;
+    if (events & FileRemoved)  flags |= IN_DELETE;
+    if (events & FileModified) flags |= (IN_MODIFY | IN_ATTRIB);
+
+    // Create watcher
+    int handle = inotify_add_watch(m_inotify, fh.path().c_str(), flags);
+    if (handle < 0) {
+        return;
+    }
+
+    // Watch directories recursively
+    if (fh.isDirectory() && recursive == Recursive) {
+        // List directory entries
+        for (auto it = fh.begin(); it != fh.end(); ++it)
+        {
+            // Check if entry is a directory
+            FileHandle fh2 = fh.open(*it);
+            if (fh2.isDirectory()) {
+                // Watch directory
+                add(fh2, events, recursive);
+            }
+        }
+    }
+
+    // Associate watcher handle with file handle
+    m_watchers[handle].fileHandle = fh;
+    m_watchers[handle].events     = events;
+    m_watchers[handle].recursive  = recursive;
+}
+
+void LocalFileWatcher::watch(int timeout)
+{
+    // Create buffer for receiving events
+    size_t bufSize = 64 * (sizeof(inotify_event) + NAME_MAX);
+    std::vector<char> buffer;
+    buffer.resize(bufSize);
+
+    // Set timeout
+    if (timeout >= 0) {
+        // Create file descriptor set
+        fd_set set;
+        FD_ZERO(&set);
+        FD_SET(m_inotify, &set);
+
+        // Convert timeout into timeval struct
+        struct timeval tval;
+        tval.tv_sec = timeout / 1000;
+        tval.tv_usec = (timeout - tval.tv_sec * 1000) * 1000;
+
+        // Set timeout on file descriptor
+        int rv = select(m_inotify + 1, &set, NULL, NULL, &tval);
+        if (rv <= 0) {
+            return;
+        }
+    }
+
+    // Read events
+    int size = read(m_inotify, buffer.data(), bufSize);
+    if (size < 0) {
+        return;
+    }
+
+    // Process all events
+    int i = 0;
+    while (i < size) {
+        // Get event
+        auto * event = reinterpret_cast<inotify_event *>(&buffer.data()[i]);
+        if (event->len) {
+            // Get event
+            FileEvent eventType = (FileEvent)0;
+                 if (event->mask & IN_CREATE) eventType = FileCreated;
+            else if (event->mask & IN_DELETE) eventType = FileRemoved;
+            else if (event->mask & IN_MODIFY) eventType = FileModified;
+            else if (event->mask & IN_ATTRIB) eventType = FileModified;
```

Comment:
I think we should add a new type FileAttribChanged instead of implicitly merging them.

Reply:
Good idea.